server/gitrest/Dockerfile

```diff
@@ -2,37 +2,83 @@
+FROM node:18.17.1-buster-slim@sha256:cb2a746612c2564d3bd0f871174618337af9d0f4d895df6a623fd7a69ca6e5bd AS runnerbase
+
+ARG SETVERSION_VERSION=dev
+ENV SETVERSION_VERSION=$SETVERSION_VERSION
+
+ARG SETVERSION_CODEVERSION=dev
+ENV SETVERSION_CODEVERSION=$SETVERSION_CODEVERSION
+
+ARG INTERDEPENDENCY_RANGE=^
+ENV INTERDEPENDENCY_RANGE=$INTERDEPENDENCY_RANGE
+
+ARG RELEASE_GROUP=gitrest
+ENV RELEASE_GROUP=$RELEASE_GROUP
+
+ARG BUILD_ROOT=/usr/FluidFramework
+ARG GITREST_ROOT=$BUILD_ROOT/server/gitrest
 
 # Add Tini
-ENV TINI_VERSION v0.18.0
+ENV TINI_VERSION=v0.18.0
 ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
 RUN chmod +x /tini
 
-# Copy over and build the server
-WORKDIR /home/node/server
+FROM runnerbase AS base
+
+# Copy over and build the server. We use the same directory structure as outside of the docker container to ensure flub works smoothly.
+WORKDIR $GITREST_ROOT
 
 COPY package*.json ./
 COPY pnpm*.yaml ./
+COPY .npmrc ./
+COPY lerna.json ./
 COPY scripts/*.* ./scripts/
 COPY .npmrc ./
 
 COPY packages/gitrest/package*.json packages/gitrest/
 COPY packages/gitrest-base/package*.json packages/gitrest-base/
 
+# Configuring PNPM_HOME allows reusing the host container's pnpm cache by mounting it to
+# this location later on in the build.
+ENV PNPM_HOME="/pnpm"
+# Add pnpm executables as well as package dependency's executables to the PATH.
+ENV PATH="$PNPM_HOME:$GITREST_ROOT/node_modules/.bin:$PATH"
 RUN corepack enable
 
 # Need to set the --unsafe-perm flag since we are doing the install as root. Consider adding an 'app' account so we
 # can do the install as node but then switch to 'app' to run. As app we won't be able to write to installed files
 # and be able to change them.
-# TODO: AB#26779: This Workaround (passing --no-frozen-lockfile) should be removed when possible.
-RUN pnpm install --unsafe-perm --no-frozen-lockfile
+RUN --mount=type=cache,id=pnpm,target=/pnpm/store\
+    pnpm install --unsafe-perm
 
 COPY . .
-RUN npm run build
+
+# Copy over fluid config files to allow flub versioning to work
+WORKDIR $BUILD_ROOT
+COPY --from=root fluidBuild.config.cjs ./fluidBuild.base.config.cjs
+COPY --from=root ./server/gitrest/fluidBuild.docker.config.cjs fluidBuild.config.cjs
+WORKDIR $GITREST_ROOT
+ENV _FLUID_ROOT_=$BUILD_ROOT
+COPY --from=root /scripts/update-package-version.sh ./scripts/update-package-version.sh
+RUN chmod +x ./scripts/update-package-version.sh
+RUN set -eu;\
+if [ "$SETVERSION_VERSION" != "dev" ];\
+    then echo "Setting package version: $SETVERSION_VERSION";\
+    ./scripts/update-package-version.sh;\
+else\
+    echo "Skipping package version for dev build.";\
+fi
+
+RUN pnpm run build
 
 # Build that actually runs
-FROM base AS runner
+FROM runnerbase AS runner
+
+WORKDIR /home/node/server
+
+COPY --from=base $GITREST_ROOT/node_modules ./node_modules
+COPY --from=base $GITREST_ROOT/packages ./packages
```

Comment:
Opportunistically reducing the number of layers of the final container by selectively copying output from `base` instead of building on top of it? I like it, just making sure if that was the intention.

Reply:
Yeah, less layers and less overall content in the aggregate. It's not super pruned as I noted in the PR description, but should be a tad better than before.